demos/slack-alert-app/go/slackbot.go

```diff
@@ -0,0 +1,166 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"strings"
+	"time"
+	"github.com/slack-go/slack"
+	"go.withpixie.dev/pixie/src/api/go/pxapi"
+	"go.withpixie.dev/pixie/src/api/go/pxapi/types"
+)
+
+var (
+	// Slack channel for Slackbot to post in.
+	// Slack App must be a member of this channel.
+	slackChannel = "#pixie-alerts"
+
+	// This PxL script ouputs a table of the HTTP total requests count and
+	// HTTP error (>4xxx) count for each service in the `px-sock-shop` namespace.
+	pxlScript = `
+import px
+
+df = px.DataFrame(table='http_events', start_time='-5m')
+
+# Add column for HTTP response status errors.
+df.error = df.http_resp_status >= 400
+
+# Add columns for service, namespace info
+df.namespace = df.ctx['namespace']
+df.service = df.ctx['service']
+
+# Filter for px-sock-shop namespace only.
+df = df[df.namespace == 'px-sock-shop']
+
+# Group HTTP events by service, counting errors and total HTTP events.
+df = df.groupby(['service']).agg(
+    error_count=('error', px.sum),
+    total_requests=('http_resp_status', px.count)
+)
+
+px.display(df, "http_table")
+`
+)
+
+func main() {
+
+	// The slackbot requires the following configs, which are specified
+	// using environment variables. For directions on how to find these
+	// config values, see: https://docs.pixielabs.ai/tutorials/slackbot-alert
+	pixieApiKey, ok := os.LookupEnv("PIXIE_API_KEY")
+	if !ok {
+		panic("Please set PIXIE_API_KEY environment variable.")
+	}
+
+	pixieClusterID, ok := os.LookupEnv("PIXIE_CLUSTER_ID")
+	if !ok {
+		panic("Please set PIXIE_CLUSTER_ID environment variable.")
+	}
+
+    slackToken, ok := os.LookupEnv("SLACK_BOT_TOKEN")
+	if !ok {
+		panic("Please set SLACK_BOT_TOKEN environment variable.")
+	}
+
+	ctx := context.Background()
+	pixieClient, err := pxapi.NewClient(ctx, pxapi.WithAPIKey(pixieApiKey))
+	if err != nil {
+		panic(err)
+	}
+	vz, err := pixieClient.NewVizierClient(ctx, pixieClusterID)
+	if err != nil {
+		panic(err)
+	}
+
+	slackClient := slack.New(slackToken)
+
+	// Send tick every 5 minutes.
+	ticker := time.NewTicker(5*time.Minute)
+	for {
+		tm := &tableMux{tables: make(map[string]*tableCollector)}
+		log.Println("Executing PxL script.")
+		resultSet, err := vz.ExecuteScript(ctx, pxlScript, tm)
+		if err != nil && err != io.EOF {
+			panic(err)
+		}
+
+		log.Println("Stream PxL script results.")
+		if err := resultSet.Stream(); err != nil {
+			fmt.Printf("Got error : %+v, while streaming.\n", err)
+		}
+
+		// Get slack message constructed from table data.
+		table := tm.GetTable("http_table").GetTableDataSync()
+		log.Println("Sending slack message.")
+		_, _, err = slackClient.PostMessage(slackChannel, slack.MsgOptionText(table, false), slack.MsgOptionAsUser(true))
+		if err != nil {
+			log.Println("Error sending to slack: " + err.Error())
+		}
+
+		// wait for next tick
+		<-ticker.C
```

Comment:
Have you checked if ctrl+c or ctrl+z will successfully exist this program? If not you may need to listen to those signals (instructions like https://golangcode.com/handle-ctrl-c-exit-in-terminal/) and do something like this (adapted from https://stackoverflow.com/questions/16466320/is-there-a-way-to-do-repetitive-tasks-at-intervals)

```
quitCh := make(chan struct{})
setupSignalHandler(quitCh)

executeTimer := time.NewTicker(5 * time.Second)
go func() {
    for {
       select {
        case <- ticker.C:
            handleTable(....)
        case <- quitCh:
            ticker.Stop()
            return
        }
    }
 }()
```

then have the ctrl+c/ctrl+z signal listener be responsible for calling the quit channel

Reply:
Yes, ctrl+c successfully exists.

Reply:
okay ignore the above remark then